src/set_theory/surreal.lean

```diff
@@ -286,12 +456,219 @@ instance : ordered_add_comm_group surreal :=
+begin
+  cases m with m m,
+  { have := nonneg_of_pos_nsmul m hx,
+    simpa only [int.of_nat_eq_coe, gsmul_coe_nat] },
+  { exfalso,
+    rwa [← int.neg_succ_not_nonneg m] }
+end
+
+lemma pos_of_pos_gsmul {m : ℤ} {x : surreal} (hm : 0 < m) (hx : 0 < x) : 0 < m • x :=
+begin
+  cases m with m m,
+  { simp at *,
+    apply pos_of_pos_nsmul hm hx },
+  { exfalso,
+    rwa [← int.neg_succ_not_pos m] }
+end
+
+lemma lt_of_gsmul_pos_lt {m : ℤ} {x y : surreal} (hm : 0 < m) (hxy : x < y) : m • x < m • y :=
+begin
+  cases m with m m,
+  { simp at *,
+    exact lt_of_nsmul_pos_lt hm hxy },
+  { exfalso,
+    rwa [← int.neg_succ_not_pos m] }
+end
+
+lemma gmul_cancel_of_pos {m : ℤ} {x y : surreal} (hm : 0 < m) (hmxy : m • x  = m • y) : x = y :=
+begin
+    contrapose hmxy,
+    cases (@ne_iff_lt_or_gt _ _ x y).1 hmxy with hmxy' hmxy',
+    { apply _root_.ne_of_lt,
+      exact lt_of_gsmul_pos_lt hm hmxy' },
+    { apply ne_of_gt,
+      change y < x at hmxy',
+      exact lt_of_gsmul_pos_lt hm hmxy' }
+end
+
+-- TODO: move this to data.int.basic
+lemma int.exists_nat_eq_of_nonneg {x : ℤ} (h : 0 ≤ x) : ∃ (y : ℕ), (y : ℤ) = x :=
+begin
+  cases x,
+  { simp only [int.of_nat_eq_coe, exists_apply_eq_apply] },
+  { refine absurd h _,
+    simp only [int.neg_succ_not_nonneg, not_false_iff] }
+end
+
+-- TODO: move this to group_theory.submonoid.membership
+lemma mem_powers_iff {α : Type*} [monoid α] (z x : α) :
+  x ∈ submonoid.powers z ↔ ∃ n : ℕ, z ^ n = x := iff.rfl
+
+-- TODO: move this to data.int
+lemma int.pow_right_injective {x : ℤ} (h : 2 ≤ x) : function.injective (λ (n : ℕ), x ^ n) :=
+begin
+  intros n m hnm,
+  obtain ⟨y, rfl⟩ : ∃ (y : ℕ), (y : ℤ) = x := int.exists_nat_eq_of_nonneg ((zero_le_two).trans h),
+  have : 2 ≤ y,
+  { rw ←int.coe_nat_le,
+    simpa using h },
+  apply nat.pow_right_injective this,
+  simpa [←int.coe_nat_pow, int.coe_nat_inj'] using hnm
+end
+
+/-- The surreal number `half`. -/
+def half : surreal := ⟦⟨pgame.half, pgame.numeric_half⟩⟧
+
+/-- Powers of the surreal number `half`. -/
+def pow_half (n : ℕ) : surreal := ⟦⟨pgame.pow_half n, pgame.numeric_pow_half⟩⟧
+
+@[simp] lemma pow_half_zero : pow_half 0 = 1 := rfl
+
+@[simp] lemma pow_half_one : pow_half 1 = half := rfl
+
+@[simp] theorem add_half_self_eq_one : half + half = 1 :=
+quotient.sound pgame.add_half_self_equiv_one
+
+lemma double_pow_half_succ_eq_pow_half (n : ℕ) : 2 • pow_half n.succ = pow_half n :=
+begin
+  have : 2 • pow_half n.succ = pow_half n.succ + pow_half n.succ, by abel,
+  rw this,
+  apply quotient.sound,
+  apply pgame.add_pow_half_succ_self_eq_pow_half,
+end
+
+/-- Exponentiation map from natural numbers to powers of integers. -/
+def pow (a : ℤ) (n : ℕ) : submonoid.powers a := ⟨a ^ n, n, rfl⟩
+
+/-- Logarithms from powers of integers to natural numbers. -/
+noncomputable def log {a : ℤ} (p : submonoid.powers a) : ℕ :=
+classical.some $ (mem_powers_iff a p.val).1 p.prop
```

Comment:
Can you use `nat.find` here instead, to make this computable?

Reply:
It works! Amazing, now everything is computable. Thanks a lot!!